dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/ElasticListener.java

```diff
@@ -349,6 +357,26 @@ public static JsonNode dockstoreEntryToElasticSearchObject(final Entry entry) th
         return jsonNode;
     }
 
+    private static Set<Partner> getPartnersWithExecutionMetrics(Set<Version> workflowVersions) {
+        return getPartnersWithMetrics(workflowVersions, EXECUTION);
+    }
+
+    private static Set<Partner> getPartnersWithValidationMetrics(Set<Version> workflowVersions) {
+        return getPartnersWithMetrics(workflowVersions, VALIDATION);
+    }
+
+    private static Set<Partner> getPartnersWithMetrics(Set<Version> versions, MetricsType metricsType) {
+        return versions.stream()
+            .filter(WorkflowVersion.class::isInstance)
+            .map(WorkflowVersion.class::cast)
+            .map(WorkflowVersion::getMetricsByPlatform)
```

Comment:
Alas, I suspect that this call is going to cause a similar performance problem to what we experienced when we retrieved all of the `SourceFiles` (before we optimized that away).   Because since the fetchtype is lazy, each `getMetricsByPlatform` call is going to trigger a separate query for each version, and thus the N+1 problem!

Some alternatives would be to:
1) "preaggregate" the metrics info in each Version, similar to how the "verified" data is currently, and work with the preaggregated data (which is retrieved eagerly, avoiding the N+1 issue).
2) write a custom query.
However, not sure how performant Option 2 will be for the huge workflows with lots of metrics data.

Reply:
I'll redo; I had thought this was similar to other iterations over versions we were already doing and we had deemed OK -- which is why I had removed the TODO statement on line 325. But this one is different, given the extra table and the potential for huge amounts of data.

Reply:
Another [more radical] idea would be to only ES index information in the current default version, thus dispensing with the need to aggregate information across all versions...

Reply:
> Another [more radical] idea would be to only ES index the current default version, thus dispensing with the need to aggregate information across all versions, for the sake of indexing, at least...

I think we should get them for all versions, especially now, when we have so few entries with metrics. Maybe that will change if we get a flood of data.

Reply:
Update, I did some experiments:

* The PR as it is now takes almost twice the time to reindex a prod snapshot, from ~130s to ~310 seconds, so yes, it probably needs to be addressed.
* The current PR, adding one line, `@BatchSize(size=25)` before `metricsByPlatform` in Version.java, caused the reindex to be almost the same, 130s vs 140s. That's borderline acceptable.
* However deleting a branch from the warp repo caused a noticeable delay from 5s to 20s. I don't think this is acceptable.
* Doing JPA queries to fetch the metrics improved performance a little, but not a lot (much less than `@BatchSize` did).

It seems like pre-aggregation might be best, we can't save the pre-aggregation to Version, because Version might be frozen. These are the options I can think of:

1. Add pre-aggregated metrics to VersionMetadata, keeping  `metricsByPlatform` in Version.
2. Add pre-aggregated metrics to VersionMetadata, moving `metricsByPlatform` from Version to VersionMetadata

I'm going to experiment with the JPA queries a little bit more.

But @kathy-t , any thoughts/recollections on why metrics is in Version and not on VersionMetadata?

Reply:
> * The current PR, adding one line, `@BatchSize(size=25)` before `metricsByPlatform` in Version.java, caused the reindex to be almost the same, 130s vs 140s. That's borderline acceptable.

FWIW, this result made me realize that I'd harbored a misconception about how `@BatchSize` works.  See
 
https://winterbe.com/posts/2012/09/07/tweak-hibernate-lazy-loading-batchsize/

Basically, given a property `bar` of `Foo` that is annotated with `BatchSize(N)`, when our code accesses `bar` in a particular `Foo`, Hibernate will run a query that retrieves `bar`, but also retrieves `bar` for N-1 other `Foos` in the current session for which `bar` is uninitialized.  Which is very cool, and not what I thought it did, which was to retrieve up to N elements of a particular `Foo`'s `bar` per query.

Reply:
> But @kathy-t , any thoughts/recollections on why metrics is in Version and not on VersionMetadata?

Hmm, I don't think I really considered putting it into VersionMetadata. I just tested on QA and you can PUT aggregated metrics for a frozen version. 

It sounds like `metricsByPlatform` should've been in VersionMetadata to begin with if the intent is that Version should be unmodifiable if it's frozen, so I would be okay with moving it to VersionMetadata if that makes the most sense.

Reply:
> > But @kathy-t , any thoughts/recollections on why metrics is in Version and not on VersionMetadata?
> 
> Hmm, I don't think I really considered putting it into VersionMetadata. I just tested on QA and you can PUT aggregated metrics for a frozen version.
> 
> It sounds like `metricsByPlatform` should've been in VersionMetadata to begin with if the intent is that Version should be unmodifiable if it's frozen, so I would be okay with moving it to VersionMetadata if that makes the most sense.

I went with a different approach, not storing pre-aggregated metrics, so I'm going to leave it as is for now.